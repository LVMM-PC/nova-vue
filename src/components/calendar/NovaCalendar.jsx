import dayjs from 'dayjs';
import Inventory from '@/utils/inventory';
import Calendar from '@/utils/calendar';
import locale from '@/mixin/locale';
import Month from './Month.jsx';

export default {
  name: 'NovaCalendar',
  components: {
    Month
  },
  mixins: [locale],
  model: {
    event: 'update'
  },
  provide() {
    return {
      NovaCalendar: this
    };
  },
  props: {
    prefixedClass: {
      type: String,
      default: `${Inventory.prefix}-calendar`
    },
    value: {
      type: Date,
      default: null
    },
    disabledDate: {
      type: Function,
      default: () => {
        return false;
      }
    },
    disabledMonthPrev: {
      type: Function,
      default: () => {
        return false;
      }
    },
    disabledMonthNext: {
      type: Function,
      default: () => {
        return false;
      }
    },
    monthSize: {
      type: Number,
      default: 1
    }
  },
  data() {
    let defaultDate = dayjs();
    if (this.value) {
      defaultDate = dayjs(this.value);
    }
    const first = Calendar.getFirstDateMomentOfMonth(defaultDate);
    this.$emit('update', first.toDate());
    return {
      weeks: Calendar.weeks,
      panelMoment: first,
      defaultFormat: Calendar.defaultFormat
    };
  },
  computed: {
    showMonthSize() {
      return this.monthSize;
    }
  },
  methods: {
    updateShowDate(date) {
      if (!date) {
        return;
      }
      if (dayjs(date).isValid()) {
        this.panelMoment = Calendar.getFirstDateMomentOfMonth(date);
      }

      this.refreshDateList();
      this.$emit('update', this.panelMoment.toDate());
      this.$emit('panelChange', this.panelMoment.toDate());
    },
    refreshDateList() {
      const $months = this.$refs['month'];
      if (!$months) {
        return;
      }
      $months.forEach($month => {
        $month.refreshDateList();
      });
    }
  },
  render() {
    const {
      $attrs,
      $listeners,
      $scopedSlots,
      novaLocale,
      prefixedClass,
      showMonthSize
    } = this;

    const calendarProps = {
      class: prefixedClass,
      attrs: {
        ...$attrs
      },
      on: {
        ...$listeners
      }
    };

    let monthScopedSlots = {};
    if ($scopedSlots.dateCellRender) {
      monthScopedSlots = {
        dateCellRender: slotProps => {
          return $scopedSlots.dateCellRender(slotProps);
        }
      };
    }

    const monthList = Array(showMonthSize)
      .fill(null)
      .map((empty, index) => {
        const monthProps = {
          key: index,
          props: {
            novaLocale,
            offset: index
          },
          ref: 'month',
          refInFor: true,
          scopedSlots: monthScopedSlots
        };

        return <Month {...monthProps}></Month>;
      });

    return (
      <div {...calendarProps}>
        <div class={`${prefixedClass}-months`}>{monthList}</div>
      </div>
    );
  }
};
