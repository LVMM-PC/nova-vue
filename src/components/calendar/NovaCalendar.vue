<template>
  <div class="nova-ui-calendar" v-bind="$attrs" v-on="$listeners">
    <div class="nova-ui-calendar-months">
      <Month
        ref="monthRef"
        :nova-locale="novaLocale"
        v-for="(month, monthIndex) in showMonthSize"
        :offset="monthIndex"
        :key="monthIndex"
      >
        <template slot="dateCellRender" slot-scope="scope">
          <slot
            name="dateCellRender"
            :date="scope.date"
            :index="scope.index"
            :offset="scope.offset"
            :paneDate="scope.paneDate"
          ></slot>
        </template>
      </Month>
    </div>
  </div>
</template>

<script>
import moment from 'moment';
import Calendar from '@/utils/calendar';
import locale from '@/mixin/locale';
import Month from '@/components/calendar/Month';

export default {
  name: 'NovaCalendar',
  mixins: [locale],
  model: {
    event: 'update'
  },
  components: {
    Month
  },
  provide() {
    return {
      NovaCalendar: this
    };
  },
  props: {
    value: {},
    disabledDate: {
      type: Function,
      default: () => {
        return false;
      }
    },
    disabledMonthPrev: {
      type: Function,
      default: () => {
        return false;
      }
    },
    disabledMonthNext: {
      type: Function,
      default: () => {
        return false;
      }
    },
    monthSize: {
      type: Number,
      default: 1
    }
  },
  data() {
    let defaultDate = moment();
    if (this.value) {
      defaultDate = moment(this.value);
    }
    let first = Calendar.getFirstDateMomentOfMonth(defaultDate);
    this.$emit('update', first.toDate());
    return {
      weeks: Calendar.weeks,
      paneMoment: first,
      defaultFormat: Calendar.defaultFormat
    };
  },
  computed: {
    showMonthSize() {
      return this.monthSize;
    }
  },
  methods: {
    updateShowDate(date) {
      if (!date) {
        return;
      }
      if (moment.isDate(date)) {
        this.paneMoment = Calendar.getFirstDateMomentOfMonth(date);
      }

      this.refreshDateList();
      this.$emit('update', this.paneMoment.toDate());
      this.$emit('panelChange', this.paneMoment.toDate());
    },
    refreshDateList() {
      let monthRef = this.$refs['monthRef'];
      if (!monthRef) {
        return;
      }
      monthRef.forEach($month => {
        $month.refreshDateList();
      });
    }
  }
};
</script>

<style lang="less">
@import '../../styles/var';

.nova-ui-calendar {
  width: 1196px;
  font-family: @font-family;
  font-size: 12px;
  background-color: #fff;
}
</style>
