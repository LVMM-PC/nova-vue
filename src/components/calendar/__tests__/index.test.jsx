import mountTest from '../../../../tests/shared/mountTest';
import NovaCalendar from '@/components/calendar/NovaCalendar';
import { mount } from '@vue/test-utils';
import zhTW from '@/locales/lang/zh-TW';
import dayjs from 'dayjs';

describe('NovaCalendar.vue', () => {
  mountTest(NovaCalendar);

  it('Calendar should support locale', function() {
    const wrapper = mount({
      data() {
        return { date: dayjs('2000-01-01').toDate() };
      },
      render() {
        return <NovaCalendar vModel={this.date} locale={zhTW} />;
      }
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('should trigger panelChange when click next or prev', () => {
    const onPanelChange = jest.fn();
    const wrapper = mount({
      data() {
        return { date: dayjs('2000-01-01').toDate() };
      },
      render() {
        return <NovaCalendar onPanelChange={onPanelChange}></NovaCalendar>;
      }
    });
    wrapper.find('.nova-calendar-next').trigger('click');
    expect(onPanelChange).toHaveBeenCalledTimes(1);
    wrapper.find('.nova-calendar-prev').trigger('click');
    expect(onPanelChange).toHaveBeenCalledTimes(2);
  });

  it('dateCellRender', () => {
    const wrapper = mount({
      data() {
        return { date: dayjs('2000-01-01') };
      },
      render() {
        return (
          <NovaCalendar>
            <div slot="dateCellRender" class="bamboo">
              Light
            </div>
          </NovaCalendar>
        );
      }
    });
    expect(
      wrapper
        .findAll('.bamboo')
        .at(0)
        .text()
    ).toEqual('Light');
  });
});
