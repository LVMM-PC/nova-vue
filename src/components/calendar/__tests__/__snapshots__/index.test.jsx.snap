// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NovaCalendar.vue Calendar should support locale 1`] = `
<div class="nova-calendar">
  <div class="nova-calendar-months">
    <div class="nova-calendar-month">
      <div class="nova-calendar-header">
        <div class="nova-calendar-weeks">
          <div class="nova-calendar-week nova-calendar-sun">
            日
          </div>
          <div class="nova-calendar-week nova-calendar-mon">
            一
          </div>
          <div class="nova-calendar-week nova-calendar-tue">
            二
          </div>
          <div class="nova-calendar-week nova-calendar-wed">
            三
          </div>
          <div class="nova-calendar-week nova-calendar-thu">
            四
          </div>
          <div class="nova-calendar-week nova-calendar-fri">
            五
          </div>
          <div class="nova-calendar-week nova-calendar-sat">
            六
          </div>
        </div>
      </div>
      <div class="nova-calendar-content">
        <div class="nova-calendar-dates">
          <div class="nova-calendar-date is-prev">
            <div class="nova-calendar-date-number">
              26
            </div>
          </div>
          <div class="nova-calendar-date is-prev">
            <div class="nova-calendar-date-number">
              27
            </div>
          </div>
          <div class="nova-calendar-date is-prev">
            <div class="nova-calendar-date-number">
              28
            </div>
          </div>
          <div class="nova-calendar-date is-prev">
            <div class="nova-calendar-date-number">
              29
            </div>
          </div>
          <div class="nova-calendar-date is-prev">
            <div class="nova-calendar-date-number">
              30
            </div>
          </div>
          <div class="nova-calendar-date is-prev">
            <div class="nova-calendar-date-number">
              31
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              1
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              2
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              3
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              4
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              5
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              6
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              7
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              8
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              9
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              10
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              11
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              12
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              13
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              14
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              15
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              16
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              17
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              18
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              19
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              20
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              21
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              22
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              23
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              24
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              25
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              26
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              27
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              28
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              29
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              30
            </div>
          </div>
          <div class="nova-calendar-date">
            <div class="nova-calendar-date-number">
              31
            </div>
          </div>
          <div class="nova-calendar-date is-next">
            <div class="nova-calendar-date-number">
              1
            </div>
          </div>
          <div class="nova-calendar-date is-next">
            <div class="nova-calendar-date-number">
              2
            </div>
          </div>
          <div class="nova-calendar-date is-next">
            <div class="nova-calendar-date-number">
              3
            </div>
          </div>
          <div class="nova-calendar-date is-next">
            <div class="nova-calendar-date-number">
              4
            </div>
          </div>
          <div class="nova-calendar-date is-next">
            <div class="nova-calendar-date-number">
              5
            </div>
          </div>
        </div>
      </div>
      <div class="nova-calendar-sidebar">
        <div title="上個月" class="nova-calendar-prev">
          上個月
        </div>
        <div class="nova-calendar-title"><span class="nova-calendar-title-support"></span> <span class="nova-calendar-title-text">2000年
1月</span></div>
        <div title="下個月" class="nova-calendar-next">
          下個月
        </div>
      </div>
    </div>
  </div>
</div>
`;
